src/V3Delayed.cpp

```diff
@@ -221,448 +293,520 @@ class DelayedVisitor final : public VNVisitor {
-            const bool partial = m_vscpAux(vscp).hasPartialUpdate;
-
-            // If partial updates are required, construct the mask and the widened value
-            AstNodeExpr* valuep = record.rhsp;
-            AstNodeExpr* maskp = nullptr;
-            if (partial) {
-                // Type of array element
-                AstNodeDType* const eDTypep = [&]() -> AstNodeDType* {
-                    AstNodeDType* dtypep = vscp->dtypep()->skipRefp();
-                    while (AstUnpackArrayDType* const ap = VN_CAST(dtypep, UnpackArrayDType)) {
-                        dtypep = ap->subDTypep()->skipRefp();
+        // Base name suffix for signals constructed below
+        const std::string baseName{"__" + vscp->varp()->shortName() + "__v"
+                                   + std::to_string(vscpInfo.nTmp++)};
+
+        // Unlink and capture the RHS value
+        AstNodeExpr* const capturedRhsp
+            = captureVal(scopep, nodep, nodep->rhsp()->unlinkFrBack(), "__VdlyVal" + baseName);
+
+        // Unlink and capture the LHS reference
+        AstNodeExpr* const capturedLhsp
+            = captureLhs(scopep, nodep, nodep->lhsp()->unlinkFrBack(), baseName);
+
+        // RHS value (can be widened/masked iff Partial)
+        AstNodeExpr* valuep = capturedRhsp;
+        // RHS mask (iff Partial)
+        AstNodeExpr* maskp = nullptr;
+
+        // Running node for LHS deconstruction
+        AstNodeExpr* lhsNodep = capturedLhsp;
+
+        // If partial updates are required, construct the mask and the widened value
+        if VL_CONSTEXPR_CXX17 (Partial) {
+            // Type of array element
+            AstNodeDType* const eDTypep = [&]() -> AstNodeDType* {
+                AstNodeDType* dtypep = vscp->dtypep()->skipRefp();
+                while (AstUnpackArrayDType* const ap = VN_CAST(dtypep, UnpackArrayDType)) {
+                    dtypep = ap->subDTypep()->skipRefp();
+                }
+                return dtypep;
+            }();
+
+            if (AstSel* const lSelp = VN_CAST(lhsNodep, Sel)) {
+                // This is a partial assignment.
+                // Need to create a mask and widen the value to element size.
+                lhsNodep = lSelp->fromp();
+                AstNodeExpr* const sLsbp = lSelp->lsbp();
+                AstConst* const sWidthp = VN_AS(lSelp->widthp(), Const);
+
+                // Create mask value
+                maskp = [&]() -> AstNodeExpr* {
+                    // Constant mask we can compute here
+                    if (AstConst* const cLsbp = VN_CAST(sLsbp, Const)) {
+                        AstConst* const cp = new AstConst{flp, AstConst::DTyped{}, eDTypep};
+                        cp->num().setAllBits0();
+                        const int lsb = cLsbp->toSInt();
+                        const int msb = lsb + sWidthp->toSInt() - 1;
+                        for (int bit = lsb; bit <= msb; ++bit) cp->num().setBit(bit, '1');
```

Comment:
Please update V3Number::setMask to accept a starting lsb (default = 0), and call here.  Please make a prepull that adds that and fixes other spots in code to use it, you can merge the prepull once passes.

    src/V3Delayed.cpp:615:                        for (int bit = lsb; bit <= msb; ++bit) cp->num().setBit(bit, '1');
    src/V3Expand.cpp:554:            for (int bit = lsb; bit < (msb + 1); bit++) maskset.setBit(bit, 1);
    src/V3Expand.cpp:657:                for (int bit = 0; bit < lhsp->widthConst(); bit++) maskwidth.setBit(bit, 1);